lib/autoupdate/agent.go

```diff
@@ -65,30 +86,204 @@ type AgentUpdateStatus struct {
+}
+
+// LocalAgentUpdaterConfig specifies configuration for managing local agent auto-updates.
+type LocalAgentUpdaterConfig struct {
+	// Log contains a slog logger.
+	// Defaults to slog.Default() if nil.
+	Log *slog.Logger
+	// InsecureSkipVerify turns off TLS certificate verification.
+	InsecureSkipVerify bool
+	// DownloadTimeout is a timeout for file download requests.
+	// Defaults to no timeout.
+	DownloadTimeout time.Duration
+	// VersionsDir for installing Teleport (usually /var/lib/teleport/versions).
+	VersionsDir string
+}
+
+// AgentUpdater implements the agent-local logic for Teleport agent auto-updates.
 type AgentUpdater struct {
+	// Log contains a logger.
 	Log *slog.Logger
+	// Pool used for requests to the Teleport web API.
+	Pool *x509.CertPool
+	// InsecureSkipVerify skips TLS verification.
+	InsecureSkipVerify bool
+	// ConfigPath contains the path to the agent auto-updates configuration.
+	ConfigPath string
+	// Installer manages installations of the Teleport agent.
+	Installer AgentInstaller
+}
+
+// AgentInstaller provides an API for installing Teleport agents.
+type AgentInstaller interface {
+	// Install the Teleport agent at version from the download template.
+	Install(ctx context.Context, version, template string) error
+	// Remove the Teleport agent at version.
+	Remove(ctx context.Context, version string) error
+}
+
+// AgentUserConfig contains user overrides for installing Teleport agents.
+type AgentUserConfig struct {
+	// Proxy address, scheme and port optional.
+	// Overrides existing value if specified.
+	Proxy string
+	// Group identifier for updates (e.g., staging)
+	// Overrides existing value if specified.
+	Group string
+	// URLTemplate for the Teleport tgz download URL
+	// Overrides existing value if specified.
+	URLTemplate string
+	// ForceVersion to the specified version.
+	ForceVersion string
+}
+
+// Enable enables agent updates and attempts an initial update.
+// If the initial update succeeds, auto-updates are enabled and the configuration is persisted.
+// Otherwise, the auto-updates configuration is not changed.
+// This function is idempotent.
+func (u *AgentUpdater) Enable(ctx context.Context, userCfg AgentUserConfig) error {
+	// Read configuration from updates.yaml and override any new values passed as flags.
+	cfg, err := u.readConfig(u.ConfigPath)
+	if err != nil {
+		return trace.Errorf("failed to read update.yaml: %w", err)
+	}
+	if userCfg.Proxy != "" {
+		cfg.Spec.Proxy = userCfg.Proxy
+	}
+	if userCfg.Group != "" {
+		cfg.Spec.Group = userCfg.Group
+	}
+	if userCfg.URLTemplate != "" {
+		cfg.Spec.URLTemplate = userCfg.URLTemplate
+	}
+	cfg.Spec.Enabled = true
+	if err := validateUpdatesSpec(&cfg.Spec); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Lookup target version from the proxy.
+	addr, err := libutils.ParseAddr(cfg.Spec.Proxy)
+	if err != nil {
+		return trace.Errorf("failed to parse proxy server address: %w", err)
+	}
+
+	desiredVersion := userCfg.ForceVersion
```

Comment:
Does enabling auto-updates and forcing a specific version make sense?

I understand that we want a command to opt out of AU and run a specific version when troubleshooting agent issues. When the version is pinned, the AUs should be disbaled to avoid accidental updates during the troubleshooting session. I'm not sure what's the motivation behind allowing the users to pick a version when enabling AUs at the same time.

If the user pins a version, in the best case they will start with an out-of-date version and catch back after a week. In the worst case they will start with an out-of-support version and won't have a working agent until the updater triggers.

If there's no strong use-case behind this flag I'd recommend to remove it. We can introduce later a "set version and opt-out of AU" command to meet the RFD requirements.

Reply:
Ah, this is not intended to be version pinning. The use case I imagined is for debugging and working around inaccessible proxies. Without it, there's no way to install the agent and hook it up to a proxy unless the proxy is available. We want a mindless `teleport-update enable` to fail in that case, but I figured we wouldn't want to block users entirely.

If you aren't convinced, how about I hide it from help output? It's hard to manually test without it, as proxies aren't serving the version yet.

Reply:
Using a hidden command sounds good to me. Another alternative is using a `TELEPORT_UNSTABLE_FORCE_VERSION` environment variable.

Reply:
Regarding the "the agent should be able to enroll into AU without talking to the proxy I think it will just cause more issues for use. If the agent cannot talk to the proxy we want to fail fast. A user could set the version and this would hide the connectivity issue.

For the special case where you do want to pre-load a Teleport version without proxy connectivity, you can always set the version without enrolling into AUs.

Reply:
Hid the flag for now